sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TryEval.scala

```diff
@@ -91,7 +91,16 @@ case class TryCast(child: Expression, toType: DataType, timeZoneId: Option[Strin
     (!Cast.needsTimeZone(child.dataType, toType) || timeZoneId.isDefined)
 
   override lazy val replacement = {
-    TryEval(Cast(child, toType, timeZoneId = timeZoneId, ansiEnabled = true))
+    if (equivalentToNonAnsiCast) {
+      Cast(child, toType, timeZoneId = timeZoneId, ansiEnabled = false)
+    } else {
+      TryEval(Cast(child, toType, timeZoneId = timeZoneId, ansiEnabled = true))
+    }
+  }
+
+  private def equivalentToNonAnsiCast: Boolean = {
+    val fromType = child.dataType
+    canUpCast(fromType, toType) || fromType == StringType
```

Comment:
Should this be 

```suggestion
    canUpCast(fromType, toType) || toType == StringType
```
?

We can cast anything to a string but we can't cast arbitrary strings to arbitrary types.

Reply:
On converting a string column to other data types(except ANSI interval types) under the non-ANSI mode, Spark SQL returns null on invalid casts. 
Thus, when try_cast a string column to other data types(except ANSI interval types), we can rewrite it as CAST with ANSI off.